pkg/subscriber/processor/evaluation_events_evaluation_count_event_persister.go

```diff
@@ -483,3 +500,113 @@ func (p *evaluationCountEventPersister) upsertFeatureLastUsedInfo(
 	}
 	return nil
 }
+
+func (p *evaluationCountEventPersister) cacheUserAttributes(envEvents environmentEventMap) {
+	p.userAttributesCacheMutex.Lock()
+	defer p.userAttributesCacheMutex.Unlock()
+	for environmentId, events := range envEvents {
+		userAttributesMap := make(map[string]*userproto.UserAttribute)
+
+		if existingCache, exists := p.userAttributesCache[environmentId]; exists {
+			for _, attr := range existingCache.UserAttributes {
+				userAttributesMap[attr.Key] = &userproto.UserAttribute{
+					Key:    attr.Key,
+					Values: make([]string, len(attr.Values)),
+				}
+				copy(userAttributesMap[attr.Key].Values, attr.Values)
+			}
+		}
+
+		for _, event := range events {
+			if event.User == nil || event.User.Data == nil {
+				continue
+			}
+
+			// Extract user attributes from User.Data
+			for key, value := range event.User.Data {
+				if key == "" {
+					continue
+				}
+
+				if attr, exists := userAttributesMap[key]; exists {
+					// Check if value already exists to avoid duplicates
+					found := false
+					for _, existingValue := range attr.Values {
+						if existingValue == value {
+							found = true
+							break
+						}
+					}
+					if !found {
+						attr.Values = append(attr.Values, value)
+					}
+				} else {
+					userAttributesMap[key] = &userproto.UserAttribute{
+						Key:    key,
+						Values: []string{value},
+					}
+				}
+			}
+		}
+
+		// Convert map to slice and save to cache
+		if len(userAttributesMap) > 0 {
+			userAttributes := &userproto.UserAttributes{
+				EnvironmentId:  environmentId,
+				UserAttributes: make([]*userproto.UserAttribute, 0, len(userAttributesMap)),
+			}
+
+			for _, attr := range userAttributesMap {
+				userAttributes.UserAttributes = append(userAttributes.UserAttributes, attr)
+			}
+			p.userAttributesCache[environmentId] = userAttributes
+		}
+	}
+}
+
+func (p *evaluationCountEventPersister) writeUserAttributesCache(ctx context.Context) error {
+	ticker := time.NewTicker(time.Duration(p.evaluationCountEventPersisterConfig.WriteCacheInterval) * time.Second)
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-ticker.C:
+			p.writeUserAttributes()
+		}
+	}
+}
+
+func (p *evaluationCountEventPersister) writeUserAttributes() {
+	p.userAttributesCacheMutex.Lock()
+	defer p.userAttributesCacheMutex.Unlock()
+
+	for _, cache := range p.userAttributesCache {
+		if cache != nil && len(cache.UserAttributes) > 0 {
+			if err := p.upsertUserAttributes(cache); err != nil {
+				continue
+			}
+		}
+	}
+	// Reset the cache
+	p.userAttributesCache = make(userAttributesCache)
+}
```

Comment:
WDYT of the following approach?
If Redis is down or being restarted, if an error happens, it would delete all in-memory cache.
The example below would only delete the data if it succeeds.

```suggestion
func (p *evaluationCountEventPersister) writeUserAttributes() {
	p.userAttributesCacheMutex.Lock()
	defer p.userAttributesCacheMutex.Unlock()

	for envID, cache := range p.userAttributesCache {
		if cache != nil && len(cache.UserAttributes) > 0 {
			if err := p.upsertUserAttributes(cache); err != nil {
				p.logger.Error(
					"Failed to save user attributes, will retry next cycle",
					zap.Error(err),
					zap.String("environmentId", envID),
				)
				continue
			}
			// If successful, delete it from the cache.
			// The failed items will remain for the next attempt.
			delete(p.userAttributesCache, envID)
		}
	}
}
```